File: Sources/AndroidSwiftUI/BackHandlerView.swift
//
//  BackHandlerView.swift
//  AndroidSwiftUI
//
//  Created by Alsey Coleman Miller on 7/16/26.
//

import AndroidKit

/// Hosts a Jetpack Compose `BackHandler` to intercept the system back button.
@JavaClass("com.pureswift.swiftandroid.BackHandlerView")
open class BackHandlerView: AndroidView.View {

    @JavaMethod
    @_nonoverride public convenience init(_ context: AndroidContent.Context?, _ callback: SwiftObject?, environment: JNIEnvironment? = nil)

    @JavaMethod
    public func getCallback() -> SwiftObject?

    @JavaMethod
    open func setBackHandlerEnabled(_ enabled: Bool)
}

@JavaImplementation("com.pureswift.swiftandroid.BackHandlerView")
extension BackHandlerView {

    @JavaMethod
    func onBack() {
        callback()
    }
}

public extension BackHandlerView {

    convenience init(_ context: AndroidContent.Context, environment: JNIEnvironment? = nil, action: @escaping () -> ()) {
        let object = SwiftObject(action, environment: environment)
        self.init(context, object, environment: environment)
    }

    var callback: (() -> ()) {
        getCallback()!.valueObject().value as! (() -> ())
    }
}

File: Demo/app/src/main/java/com/pureswift/swiftandroid/BackHandlerView.kt
package com.pureswift.swiftandroid

import android.content.Context
import android.widget.FrameLayout
import androidx.activity.compose.BackHandler
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView

// Hosts a Jetpack Compose `BackHandler` to intercept the system back button and forward it to
// Swift via `onBack()`. Renders no visible UI of its own; used only for its back-press interception.
class BackHandlerView(context: Context, val callback: SwiftObject) : FrameLayout(context) {

    private var handlerEnabled by mutableStateOf(true)

    init {
        val composeView = ComposeView(context)
        composeView.setContent {
            BackHandler(enabled = handlerEnabled) {
                onBack()
            }
        }
        addView(composeView)
    }

    fun setBackHandlerEnabled(value: Boolean) {
        handlerEnabled = value
    }

    external fun onBack()
}
